---
title: 'Data Visualization 2024:Project (R)'
author: "Aman Satyendra Yadav LSI - II [22305256]"
date: "2024-07-09"
output: 
  html_document:
    number_sections: true
    toc : true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Information
Hi, I am **Aman Yadav**, a student in LSI-II at the Technical Institute of Deggendorf, matriculation number **22305265**. This HTML document presents my student project for Semester II, 2024, focusing on the Data Visualization course. The project includes a variety of plots, each accompanied by corresponding R code and brief explanations. All data used for processing, filtering, and plotting has been provided by our university lecturer. The Data consists

**1)**The Cancer Genome Atlas (TCGA): catalogs genetic mutations in over 30 cancer types using genome sequencing and bioinformatics. It         provides publicly available data on somatic mutations, copy number variations, gene expression, and clinical information from             thousands of patients, serving as a key resource for cancer research.

**2)** GENCODE : a scientific project that produces detailed annotations of human and mouse genomes. It identifies and classifies gene         features such as protein-coding genes, non-coding RNAs, and pseudogenes, providing comprehensive information on gene structures,          transcripts, and functional elements to support genomics, biology, and medical research.

**3)** Differential gene expression (DGE) analysis: it identifies genes with significant expression changes between different conditions,         such as cell types, disease states, or treatments. These genes help decipher the biological factors behind phenotypic differences         and can serve as potential biomarkers or therapeutic targets.


# Libraries

The following libraries were used to complete this project. The installation of libraries and loading them are done in two different steps. Installation only needs to be done once, but we must load the libraries every time we start coding.

```{r instalation}
options(repos = c(CRAN = "https://cloud.r-project.org/"))
install.packages(c(
  "tidyverse", 
  "ggplot2", 
  "gridExtra", 
  "readr", 
  "ggrepel", "scales",
  "ggrepel"))


```

Once installed, these libraries can be loaded anywhere in the code for plotting. Therefore, I have done the same for this project.

# TCGA expression data

The Cancer Genome Atlas (TCGA) is a comprehensive project that aims to catalog genetic mutations responsible for cancer using genome sequencing and bioinformatics. TCGA has collected and analyzed tumor and normal tissue samples from thousands of patients across more than 30 different types of cancer. The data generated by TCGA is publicly available, providing researchers with access to a rich resource for cancer research. This database includes information on somatic mutations, copy number variations, gene expression levels, and clinical data.

## Loading and preparing The TCGA data.

To further analyze the data, I downloaded the dataset to my computer and loaded it into my R environment. But before loading it in R we much load the libraries which we installed in the 1st step. Following are the libraries which is required for this task

```{r loading libraries, include=FALSE}
library(tidyverse)
library(ggplot2)
library(dplyr)
library(gridExtra)
library(readr)
library(grid)
library(scales)
library(ggrepel)
```

After loading the libraries, I can load the dataset into my R environment

```{r loading data set, echo=TRUE}
sample_file <- read.csv("C:/Users/Aman Yadav/Documents/R/tcga_cancer.csv")
```

Once completed loading the sample we can check it by using the following code

```{r viewing data, echo=TRUE}
#head(sample_file)
#View(sample_file)  
#summary(sample_file) 
```

I used the **'view'** function to inspect the dataset, and for additional clarity, I utilized the **'summary'** function. The 'summary' function provided me with a brief overview of the dataset, including the number of columns and their names. The 'head' function also comes in handy if we only want to see the first few columns instead of the entire dataset.

The next step is filtering the data. Since the dataset provided contains a large amount of information, following the instructions, I filtered out 200 samples and selected genes numbered 1 to 43.

```{r data filteration }

filtered_data <- sample_file %>%
  slice(1:200) %>%
  select(3:which(names(sample_file) == "gene_43"))

```

As mentioned in the code above, I assigned a new variable named 'filtered_data'. This variable contains the original dataset with only 200 samples and genes numbered 1 to 43. The **%\>%** pipe operator in R is highly useful for chaining operations together. In my code, I used **\<-** to assign the dataset to a new variable and then chained functions using %\>%. I utilized the slice() function to filter columns representing our samples from 1 to 200. Additionally, I employed the select() function from the dplyr package to select columns from the 3rd column up to and including 'gene_43' in the sample_file dataset. This was achieved by finding the index of the column named 'gene_43' using **which(names(sample_file) == "gene_43")**, where **which()** returns positions where the condition is TRUE. An alternative to select(3:which(names(sample_file) == "gene_43")) is **select(3:43)**, which can be used if the exact positions of the genes in the dataset are known and meet the plotting condition

# Ploting

## Expression distribution Gene1

The **"filtered_data"** contains the data set which I need for my first plotting. Visualizing the distribution of the expression values for Gene1 using a histogram.

```{r Histogram}



histogram_plot <- filtered_data %>%
  ggplot(aes(gene_1)) + 
  geom_histogram(binwidth = 0.4, fill = "grey", color = "black")+
  labs(title = "Expression of Gene_1", x = "RNA", y = "count")+ ylim(0, 50) + 
  scale_x_continuous(expand = expansion(mult = c(0.0001, 0.0001))) + 
  scale_y_continuous(expand = expansion(mult = c(0.001, 0.3))) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1.5),
        axis.text.y = element_text(size = 14),axis.text.x = element_text(size = 14))
histogram_plot

```

The histogram graph is plotted using **ggplot2** therfore I load it in the begining along with other modules to prevent any errors caused by missing libraries.

The basic structure of a ggplot plot involves specifying a dataset, a mapping of variables to aesthetics (such as the x and y axes), and one or more layers that add graphical elements.

In short **ggplot(data , mapping = (aes(x , y)))**. The **"data"** argument in ggplot will take the main dataset but instead of writing my as ggplot(data = filterd_data) I used the **%\>%** symbol so I dont have to write this everytime in the code and prevent spelling errors. In the mapping argument of ggplot we must clarify our x and y axis. I mentioned the x axis as "gene_1" which is already prresent in my main datasheet but didnt mentioned anything for y axis because I am intrested in the count of "gene_1". The y-axis is automatically set to display the count (frequency) of the occurrences of the data in each bin by default. This means you do not need to explicitly specify the y aesthetic; ggplot2 handles it for you.

The **"+"** allow us to add more functions countinuesly to the previous code. In the code above I have used geom_histogram which is a function in ggplot for ploting a histogram, Inside this function I have specified my astheic requirments for the graph for example **"geom_histogram(binwidth = 0.4, fill = "grey", color = "black")".**

Bin width decides the size(width) of each bin in the plot and hence plays a very important role. The bin width can decide how many data has to be shown in the plot. for instance if the binwidth is set very high, the histogram will have fewer bins, each covering a larger range of the data. This can have several effects on the appearance and interpretability of the histogram

Less Detail: When we use a very wide bin width, our histogram will show fewer bars, each covering a broader range of values. This means we'll miss out on the finer details in our data, making it harder to see specific patterns or identify outliers.

Overgeneralized View: With larger bins, the histogram lumps together a lot of data into each bin. This can hide important features, like peaks or multiple clusters, leading to a less accurate picture of your data’s distribution.

Since each bin is collecting more data points, the bars in our histogram will be taller. Depending on the data, this can make the histogram look disproportionate, with a few very tall bars dominating the plot.The choice of bin width can greatly affect how we interpret our data. If the bins are too wide, we might get a skewed sense of the data's shape and characteristics, possibly leading to wrong conclusions.

Here is a diffrent plot with diffrent bin widths.

```{r Histogram diff binwidth ..}


histogram_plot1 <- filtered_data %>%
  ggplot(aes(gene_1)) + 
  geom_histogram(binwidth = 2, fill = "grey", color = "black")+
  labs(title = "Expression of Gene_1", x = "RNA", y = "count")+ ylim(0, 50) + 
  scale_x_continuous(expand = expansion(mult = c(0.0001, 0.0001))) + 
  scale_y_continuous(expand = expansion(mult = c(0.001, 0.3))) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1.5),
        axis.text.y = element_text(size = 14),axis.text.x = element_text(size = 14))



histogram_plot2 <- filtered_data %>%
  ggplot(aes(gene_1)) + 
  geom_histogram(binwidth = 0.1, fill = "grey", color = "black")+
  labs(title = "Expression of Gene_1", x = "RNA", y = "count")+ ylim(0, 50) + 
  scale_x_continuous(expand = expansion(mult = c(0.0001, 0.0001))) + 
  scale_y_continuous(expand = expansion(mult = c(0.001, 0.3))) +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 1.5),
        axis.text.y = element_text(size = 14),axis.text.x = element_text(size = 14))



grid.arrange(histogram_plot1,histogram_plot2 , nrow = 1)




```

When we look at the first graph with a bin width of 2, it’s clear that the data is overly generalized. This means that most of the data is clumped into just a few bars, and we lose a lot of detail about the distribution.On the other hand, the second graph has a bin width of 0.1, which is quite small. While this graph shows the data in much more detail, with each bar representing a small range of values, it becomes cluttered and hard to interpret. It’s difficult to extract meaningful insights from this much detail.

In the line **labs(title = "Expression of Gene_1", x = "RNA", y = "count")**, the **labs** function stands for labels. It’s part of the ggplot2 package and is used to add or modify the title, x-axis label, and y-axis label of the plot.

The **ylim(0, 50)** part of the code sets the range of the y-axis. In this case, I’ve set the maximum limit to 50 to make sure the data fits well within the visible area of the plot.

The **scale_x_continuous** function is used to fine-tune the appearance and behavior of the x-axis. This can include setting the axis limits, defining where the axis ticks appear, customizing the axis labels, and adjusting the padding around the plot. Here, scale_x_continuous(expand = expansion(mult = c(0.0001, 0.0001))) was used to minimize the extra space around the x-axis, making the plot fit more snugly around the data.

Finally, I used the **theme** function from ggplot2 to customize other aspects of the plot, such as adding a border around the plot, and setting the font sizes for the axis text.

## Gene expression correlation

I used scatter plot to analyze the corelation between gene_1 and gene_12.

```{r scatter plot and correlation}
scatter_plot <- filtered_data %>%
  ggplot(aes(gene_1, gene_12)) + geom_point( color ="cyan",size = 3 , alpha = 1000,) + 
  geom_smooth(method = lm , linetype = "twodash" , alpha = 0.4) +
  labs(title = "Scatter Plot with Linear Regression Line",x = "Gene 1 Expression",y = "Gene 12 Expression") +
  theme_minimal() + annotate("text", x = max(filtered_data$gene_1), y = max(filtered_data$gene_12), 
                             label = paste("R:", round(cor(filtered_data$gene_1, filtered_data$gene_12),2)), 
                             hjust = 1, vjust = 1, size = 5, color = "black") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 4), axis.text.y = element_text(size = 14),
        axis.text.x = element_text(size = 14))

scatter_plot

```

In my analysis, I'm exploring the relationship between gene_1 and gene_12 in my dataset. To visualize this relationship and understand any potential trends, I used the geom_smooth function in R. When we use geom_smooth with the method = "lm" argument, it fits a straight line (linear model) to our data points. This is useful when we want to explore whether there’s a linear trend or relationship between two variables.

As you can see in my code i have menstioned the corelation value in labbels which allows me to display it on the top right corner along with the regression line. This modification ensures that my plot includes the correlation coefficient in a clear and informative manner, complementing the visual representation provided by geom_smooth with meaningful numerical context.

The correlation value, typically represented as **r**, quantifies the degree of linear relationship between two variables. The correlation coefficient ranges from -1 to +1, where values closer to 0 indicate weaker correlation. The correlation value provides insight into how changes in one variable are associated with changes in another. However, it’s important to note that correlation does not imply causation. A high correlation does not necessarily mean that changes in one variable cause changes in the other; it simply indicates a relationship.

In our case, the correlation value **r=−0.062** indicates a very weak negative linear relationship between the two variables. As mentioned earlier, this correlation value is close to 0, suggesting a weak relationship. The negative sign indicates inverse proportionality, meaning if one variable increases, the other variable tends to decrease.

In summary, if gene_1 is expressed more, gene_12 tends to be expressed less or negatively, based on the weak negative correlation coefficient of **R= -0.062**

The grey line which fall on the linear regression line tell us the expected area where the future data points will fall. It is wider than the confidence interval because it accounts for both the uncertainty in estimating the regression line and the variability of the data around the line. It also shows the range in which we expect the true regression line to lie with a certain level of confidence. It accounts for the uncertainty in estimating the regression line from the sample data.

## Class proportions

```{r stacked bar chart}
sample_labels <- read.csv("C:/Users/Aman Yadav/Documents/R/tcga_cancer_labels.csv")
#view(sample_labels)

genes_percentage <- sample_labels %>%
  slice(1:200) %>%
  count(Class) %>%
  mutate(percentage = n / sum(n) * 100)

# Create the stacked bar chart
stacked_bar_chart <- ggplot(genes_percentage, aes(x = "", y = percentage, fill = Class)) +
  geom_bar(stat = "identity", width = 0.2, color = "black") +  
  geom_text(aes(label = n),
            position = position_stack(vjust = 0.5),  
            size = 4, color = "black") +  
  scale_fill_manual(values = c("PRAD" = "#FFCCCC", "LUAD" = "lightblue", "KIRC" = "#CCFF99", "COAD" = "#FFCCFF", "BRCA" = "#FFCC99")) +
  labs(
    x = "",
    y = "Percentages",
    title = "Distribution of gene classes (n = 200)"
  ) + 
  coord_flip() + 
  theme_minimal() +
  theme(
    axis.title.y = element_text(margin = margin(r = 10)), 
    axis.text.y = element_text(margin = margin(r = 5)), plot.margin = unit(c(5, 5,5,5), "mm"), 
    legend.position = "top")

# Print the plot
print(stacked_bar_chart)
```

I believe that using color in stacked bar graphs is an excellent choice, as clear differentiation between categories is essential when representing data within a single bar. While patterns can also differentiate categories, they often lack aesthetic appeal. Additionally, color accessibility is important for ensuring all viewers can interpret the data, and consistent color schemes across different charts enhance comprehension. Colors can convey meaning, such as using warm colors for one category and cool colors for another, and a well-designed legend can further aid interpretation. Strategic color use can also highlight key data points, guiding the viewer's attention to important insights.


## Combining the plot

I used the gridExtra pakkage to combine all the graphs above. The pakkage allow us to put all the graphs in one image according to the size and placements.The **layout_matrix** specifies the arrangement of the plots.**rbind c(3 ,3)** means The second row contains the stacked bar chart (3), which spans both columns. The code below shows the same

```{r combining the plots}
layout_matrix <- rbind(c(1, 2),
                        c(3, 3))
grid.arrange(histogram_plot, scatter_plot, stacked_bar_chart, nrow = 2, layout_matrix = layout_matrix)

```

I used the same code to combine two different histograms with different bin widths together. The code begins with "grid.arrange" and takes the variable names containing our plots. Another argument specifies how many rows I want for organizing the graphs.

# Human genome

GENCODE is a scientific project focused on producing a comprehensive and high-quality annotation of the human and mouse genomes. It aims to identify and classify all gene features, including protein-coding genes, non-coding RNAs, and pseudogenes. The project provides detailed information on gene structures, transcripts, and their functional elements, supporting research in genomics, biology, and medicine.

## Loading and prepareing data

The downloaded file contains 20,000 randomly selected annotated genes from the GENCODE gtf file for the human genome (hg38).

I executed the following transformation steps before using the data for plotting.

**Filter for genes whose type is either “protein_coding”, “lncRNA”, “processsed_pseudogenes” or “transcribed_unprocessed_pseudogene”.**

```{r loading and filtering}
sample_4.0 <- read_tsv("C:/Users/Aman Yadav/Documents/R/hg38_genes.tsv")
#view(sample_4.0) 

filtered_genes <- c('protein_coding', 'lncRNA', 'processed_pseudogenes', 'transcribed_unprocessed_pseudogene')

filtered_gene_main <- sample_4.0 %>%
  filter(gene_type %in% filtered_genes)

```

The variable **"sample_4.0"** contains the main datasheet. I used another variable, "filtered_genes," as a placeholder for all the genes that need to be filtered. Ultimately, I filtered my main datasheet into a new variable called "filtered_gene_main" using the **%in%** symbol. The **%in%** symbol functions similarly to the **"in"** command in Python. The filter function is employed to extract important data from the main datasheet. Inside the filter function, I specified a condition for filtering, such as **(gene_type %in% filtered_genes)**.

**Rename the group of “protein_coding” genes to “mRNA” and combine “processsed_pseudogenes” plus “transcribed_unprocessed_pseudogene” to a new group called “Pseudo”.**

```{r filtering gene}

filtered_gene_main1 <- filtered_gene_main %>%
  mutate(
    modified_gene_names = case_when(
      gene_type == 'protein_coding' ~ 'mRNA',
      gene_type == 'processed_pseudogenes' ~ 'Pseudo',
      gene_type == 'transcribed_unprocessed_pseudogene' ~ 'Pseudo',
      TRUE ~ gene_type))
```

I assigned a new variable called **filtered_gene_main1**, which will contain the renamed groups. In this code, I used the **mutate** function to add another column named **modified_gene_names**. This column will contain the names that we want to assign to different gene types such as **"protein_coding", "processed_pseudogenes", "transcribed_unprocessed_pseudogene"**.

The **case_when()** function evaluates each condition sequentially and assigns a new value to modified_gene_names based on the condition that is true, for example **gene_type == "protein_coding" \~ "mRNA"**. The case_when function is very similar to the **if** function in Python.

**Keep only genes with a length (distance between start and end) higher than 1,000 and lower or equal than 20,000.**

```{r filtering gene based on lenght, include=FALSE}
filtered_gene_main %>%
  mutate(
    modified_gene_names = case_when(
      gene_type == 'protein_coding' ~ 'mRNA',
      gene_type == 'processed_pseudogenes' ~ 'Pseudo',
      gene_type == 'transcribed_unprocessed_pseudogene' ~ 'Pseudo',
      TRUE ~ gene_type)) %>%
  mutate(
    gene_lenght = (end_gene - start_gene )
  ) %>% filter(gene_lenght > 1000 & gene_lenght<= 20000)

```

I used **mutate** again to create a new column **gene_length**. This column stores the calculated gene length by subtracting **start_gene** from **end_gene** in the database. This addition makes it much easier to filter genes based on their length, although creating a new column is not strictly necessary. I did this primarily for clarity and to double-check the data for any issues. Finally, I filtered the dataset based on gene_length.

# Plotting

## Gene count

**summarizing the total count of genes per class represented by a classical bar chart. Using the newly formed categories and set an appropriate range of the scale.**

```{r barplot}
sample_4.0 <- read_tsv("C:/Users/Aman Yadav/Documents/R/hg38_genes.tsv")


filtered_genes <- c('protein_coding', 'lncRNA', 'processed_pseudogenes', 'transcribed_unprocessed_pseudogene')


filtered_gene_main <- sample_4.0 %>%
  filter(gene_type %in% filtered_genes)


filtered_gene_main1 <- filtered_gene_main %>%
  mutate(
    modified_gene_names = case_when(
      gene_type == 'protein_coding' ~ 'mRNA',
      gene_type == 'processed_pseudogenes' ~ 'Pseudo',
      gene_type == 'transcribed_unprocessed_pseudogene' ~ 'Pseudo',
      TRUE ~ gene_type)) %>%
  mutate(
    gene_lenght = (end_gene - start_gene )
  ) %>% filter(gene_lenght > 1000 & gene_lenght<= 20000)

custom_order <- c( 'Pseudo', 'mRNA','lncRNA')
filtered_gene_main1$modified_gene_names <- factor(filtered_gene_main1$modified_gene_names, levels = custom_order)


pseudo_count <- filtered_gene_main1 %>%
  count(gene_type)



colors <-c( "#ff66ff", "#ff6600", "#339933")
Bargraph_plot <- ggplot(filtered_gene_main1, aes(x = modified_gene_names, fill = modified_gene_names)) +
  geom_bar(width = 0.9, aes(color = modified_gene_names), stat = "count") +
  labs(title = "Total Gene Number", x = "Gene Type", y = "Count") + 
  scale_y_continuous(expand = expansion(mult = c(0.08, 0.08))) + 
  theme_minimal() + ylim(0, 4000) +
  scale_fill_manual(values = colors) +  
  scale_color_manual(values = colors) + 
  theme(
    panel.border = element_rect(color = "black", fill = NA, size = 1.5),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_text(angle = 0, hjust = 1)
  )


Bargraph_plot

```

he above graphs show the number of genes and their concentration in the given sample. I used **geom_bar** from the ggplot2 library to create the bar graphs. Most of the code is similar in every plot, with a few changes made. I utilized the **factor** function to organize my bars or variables on the X-axis according to the required conditions. **custom_order** is a vector that specifies the desired order of the levels for the modified_gene_names variable, listing three categories: **'lncRNA', 'Pseudo', and 'mRNA'**.

**filtered_gene_main1$modified_gene_names** refers to the column **modified_gene_names** in the filtered_gene_main1 dataframe. The levels argument of the factor() function is set to custom_order. This argument specifies the order of levels (categories) for the factor variable. By setting **levels = custom_order**, I am explicitly defining the order in which the categories will appear in plots or summaries.


## Length distribution

The density plot is a visual tool that shows the distribution of data by smoothing out jagged edges often seen in histograms. It uses a continuous curve to represent the shape and concentration of data points, making it easier to identify patterns and understand the overall distribution of a dataset.

```{r density plot, echo=TRUE}
sample_4.0 <- read_tsv("C:/Users/Aman Yadav/Documents/R/hg38_genes.tsv")


filtered_genes <- c('protein_coding', 'lncRNA', 'processed_pseudogenes', 'transcribed_unprocessed_pseudogene')


filtered_gene_main <- sample_4.0 %>%
  filter(gene_type %in% filtered_genes)


filtered_gene_main1 <- filtered_gene_main %>%
  mutate(
    modified_gene_names = case_when(
      gene_type == 'protein_coding' ~ 'mRNA',
      gene_type == 'processed_pseudogenes' ~ 'Pseudo',
      gene_type == 'transcribed_unprocessed_pseudogene' ~ 'Pseudo',
      TRUE ~ gene_type
    ),
    gene_length = (end_gene - start_gene)
  ) %>%
  filter(gene_length > 1000 & gene_length <= 20000)

custom_order <- c('Pseudo', 'mRNA', 'lncRNA')
filtered_gene_main1$modified_gene_names <- factor(filtered_gene_main1$modified_gene_names, levels = custom_order)

medians <- filtered_gene_main1 %>%
  group_by(modified_gene_names) %>%
  summarize(median_length = median(gene_length))

# Create density plot
density_plot1 <- ggplot(filtered_gene_main1, aes(x = gene_length, fill = modified_gene_names)) +
  geom_density(size = 1, alpha = 0.3) +
  ylim(0, 6e-04) + xlim(0, 20000) +
  geom_vline(data = medians, aes(xintercept = median_length, color = modified_gene_names), 
             linetype = "dashed", size = 1) +
  scale_y_continuous(labels = scientific_format(), expand = expansion(mult = c(0.001 , 0.001))) +
  labs(title = "Length distribution of gene classes",
       x = "Gene length (bp)",
       y = "Density") + 
  theme_minimal()


density_plot1
```

As you can see in the code, before plotting the density plot, I calculated the median of the given variable, which is assigned to the variable named **median**. The **group_by(modified_gene_names)** This groups the data by the modified_gene_names variable.

**summarize(median_length = median(gene_length))** Within each group defined by modified_gene_names, this calculates the median of the **gene_length column**. The result is a new dataframe (medians) where each row corresponds to a unique modified_gene_names category, and the **median_length** column contains the **median gene length** for that category.This means that rows with the same modified_gene_names value will be grouped together for the next operations.This will be used further in the code for plotting a median line in the graph. 

The ggplot2 library also includes a built-in function for plotting density plots, called **geom_density()**. This function allows us to visualize the distribution of data and supports arguments such as **size** and **alpha**, similar to other plot types. Additionally, to display median lines for the variables, I utilized **geom_vline** (where **'v'** stands for vertical), specifying the median values that were calculated beforehand. Along with that I gave some other arguments like **aes(xintercept = median_length, color = modified_gene_names), linetype = "dashed", size = 1)** for customizing the aesthetics. The **median_lenght** is the variable which holds the calculated median for each variables separately.  In this code **scale_y_continuous(labels = scientific_format()**
I used **scientific_format** from the **scale** library to change the yaxis according to the provided graph. 

As you can see, the density plot does a very good job of displaying the distribution and density of genes. However, the color gradients and overlapping lines of different variables sometimes create interpretation issues.
We can use other plots in this case for clear visuals and easy interpretation. for example a **Diamond**.


```{r Diamond}
diamond_plot <- ggplot(filtered_gene_main1, aes(x = modified_gene_names, y = gene_length, fill = modified_gene_names)) +
  geom_violin(trim = FALSE, alpha = 0.7) +
  stat_summary(fun = median, geom = "crossbar", aes(color = modified_gene_names), width = 0.5, size = 0.6)+  
  scale_y_continuous(expand = expansion(mult = c(0.05 , 0.05))) +
  labs(title = "Length Distribution of Gene Classes",
       x = "Gene Type",
       y = "Gene Length (bp)",
       fill = "Gene Type") + 
  scale_fill_manual(values = c("lightgreen", "orange", "purple")) +  
  scale_color_manual(values = c("lightgreen", "orange", "purple"))   +  
  guides(color = FALSE)  #

diamond_plot
```

The Diamond plot can give a clear illustration separately so we can interpret. While a density plot shows the shape of the distribution, a diamond plot provides a clearer visualization of the density at different values of the variable (in this case, **gene lengths**) across different categories (**gene types**). The width of each diamond represents the density of data points.  

Diamond plots are also useful when comparing distributions across multiple groups (e.g., different gene types). They provide a visual comparison of the shape and spread of distributions, making it easier to identify differences and similarities between groups.

Here is the clear comparision.

```{r comparision }
grid.arrange(density_plot1, diamond_plot, nrow = 2)
```

In conclusion, I believe the diamond plot is superior to the density plot for this task due to its ability to simultaneously display the distribution shape, density variations across categories, and summary statistics like the median. This makes it easier to compare distributions across different gene types while providing a visually compelling representation of the data's variability and central tendency.

## Combining and saving the image

```{r}
Bargraph_plot <- Bargraph_plot +
  labs(title = "A: Length Distribution by Gene Type")

density_plot1 <- density_plot1 +
  labs(title = "B: Gene Length Distribution")


combined_plot <- grid.arrange(Bargraph_plot, density_plot1, ncol = 2 ,widths = c(8, 8))


ggsave("combined_plot_with_labels.png", combined_plot, width = 5, height = 5)
```

I used the same **grid.arrange()** function to combine the two plots together and **ggsave()** to save them as one image.
**ggsave("combined_plot_with_labels.png", combined_plot, width = 10, height = 3)** this code saved my image with the label.



# Differential gene expression

Differential gene expression (DGE) analysis is a process that aims to determine genes which exhibit significant changes in expression levels between different conditions like cell types, disease states, or experimental treatments. Often, these genes are crucial in deciphering the biological factors driving phenotypic differences and can serve as potential biomarkers or therapeutic targets.

## Loading and prepareing data

The dataset contains test results for 5000 genes from a DGE analysis using the DESeq2 package. In addition, the table includes normalized RNA-seq counts for the two conditions analyzed (3 replicates each).

Loading the data 
```{r loading deseq2_datasheet}
deseq2_datasheet <- read_tsv("C:/Users/Aman Yadav/Documents/R/deseq2_dummy.txt")
```

preparing The data **selecting the top 2000 genes according to the average of the normalized count values (highest values)**

```{r preparing deseq2_datasheet }



genes_2000 <- deseq2_datasheet %>%
  arrange(desc(baseMean)) %>%
  slice(1:2000) %>%
  mutate(
    regulation = ifelse(log2FoldChange > 0 & -log10(pvalue) > 1.5, "Upregulated", 
                        ifelse(log2FoldChange < 0 & -log10(pvalue) > 1.5, "Downregulated", "Not Significant")))
```

I assinged a new variable **genes_2000** for a filtered data set. The base mean is already given in the datasheet do the **arrange(desc(baseMean))** command will arrange the rows of the data frame in descending order based on the baseMean column. baseMean typically represents the average expression level of the genes across all samples. Sorting in descending order ensures that genes with the highest expression levels are considered first.

The **slice()** reduces the data to only include the 2000 genes with the highest baseMean

I added one more columns (**regulation**) using mutate which will hold the values according to the requirement of **log2FoldChange** and **-log10(pvalue)**. The **ifelse** staement plays important role here just like **if** in python. 


## volcano plot

```{r Volcano plot}

deseq2_datasheet <- read_tsv("C:/Users/Aman Yadav/Documents/R/deseq2_dummy.txt")

genes_2000 <- deseq2_datasheet %>%
  arrange(desc(baseMean)) %>%
  slice(1:2000) %>%
  mutate(
    regulation = ifelse(log2FoldChange > 0 & -log10(pvalue) > 1.5, "Upregulated", 
                        ifelse(log2FoldChange < 0 & -log10(pvalue) > 1.5, "Downregulated", "Not Significant")))

volcano_plot <- genes_2000 %>%
  ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(aes(color = regulation)  ) +
  scale_color_manual(values = c("Downregulated" = "cyan", "Upregulated" = "maroon" ,"Not Significant" = "grey")) +
  geom_text_repel(
    data = subset(genes_2000, padj < 0.1 & log2FoldChange > 2),
    aes(label = gene_id),
    size = 3,
    box.padding = 0.3,
    point.padding = 0.5,
    segment.color = 'grey50'
  ) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "black") +
  geom_hline(yintercept = 1.5, linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(
    panel.border = element_rect(color = "black", fill = NA, size = 1.5)) +
  labs(
    title = "Volcano Plot",
    x = "(Log2) Fold Change",
    y = "Adjusted P-Value (log10)"
  )

# Print the plot
print(volcano_plot)

```

After filtering the data, the first step I performed was to plot the points. By using **geom_pont** , I set the point color based on the regulation column, so each different regulation value will have a different color. Furthermore, by using **scale_color_manual(values = c("Downregulated" = "cyan", "Upregulated" = "maroon", "Not Significant" = "grey"))**, I specified which values should be assigned which colors. **scale_color_manual** is part of the ggplot2 package. This package is widely used for creating plots and visualizations in R, offering extensive capabilities for customizing colors, scales, and aesthetics within plots. By using **geom_text_repel()** I added all the labels to all significant genes with a log2 fold change higher than 2. The **subset()** is a function that allows us to extract rows from a data frame where certain conditions are met. The condition I mentioned in the subset is **padj < 0.1 & log2FoldChange > 2** which staisfy our plotting requirement. The lines on the plot were drawn by **geom_vline** and **geom_hline**


By the help of this volcano plot, we can identify which genes exhibit significant changes in expression between experimental conditions compared to controls. Genes located towards the right side of the plot with positive log2FoldChange values and colored in maroon (Upregulated) are indicative of significant upregulation. Conversely, genes located towards the left side with negative log2FoldChange values and colored in cyan (Downregulated) suggest significant downregulation. The plot also highlights genes colored in grey (Not Significant), which indicates their expression changes are not statistically significant based on the specified thresholds. Threshold lines such as the vertical line at log2FoldChange = 0 and the horizontal line at -log10(pvalue) = 1.5 further aid in interpreting the significance and directionality of gene expression changes. Overall, the volcano plot serves as a visual aid to quickly discern and prioritize genes for further biological investigation based on their differential expression and statistical significance.


## Heatmap

```{r heatmap}
deseq2_res <- read_tsv("C:/Users/Aman Yadav/Documents/R/deseq2_dummy.txt")

deseq2_res_f <- deseq2_res %>%
  arrange(padj) %>%
  slice(1:20)

# Create the plot
heatmap1 <- deseq2_res_f %>% 
  dplyr::select(gene_id, 9:14) %>%
  pivot_longer(cols = 2:7, names_to = "condition", values_to = "counts") %>%
  ggplot(aes(x = condition, y = gene_id, fill = log2(counts))) +
  geom_tile(color = "black") +
  geom_text(aes(label = round(log2(counts), digits = 2))) +
  coord_cartesian(expand = FALSE) +
  scale_fill_distiller(palette = "Purples") +
  labs(x =  "Sample", y = "Gene IDs", fill = "Counts (log2)") +
  theme_minimal() +
  theme(
    legend.position = "top", 
    legend.ticks = element_blank(), 
    title = element_text(size = 13),
    axis.text = element_text(size = 11, colour = "black"),
    axis.title = element_text(size = 15, face = "bold"),
    panel.border = element_rect(size = 3, fill = NA)
  )

heatmap1

```



I observe that the heatmap shows differences in gene expression between cancer and normal samples, with some genes showing clear patterns. For example, **PTMA** has higher expression in cancer samples compared to normal samples, suggesting it might play a role in cancer. On the other hand, **CHAC1** has lower expression in cancer samples, which might mean it's more active in normal cells.

Additionally, genes like **SCML4** and **PLA1A** have higher expression in normal samples, while **MYL6** shows high and consistent expression in both cancer and normal samples, but slightly lower in the normal ones. **RGS3** stands out with a significant drop in expression from cancer to normal samples, hinting it could be useful for diagnosing cancer.

Other genes like **EEA1** and **CD44** are highly expressed in all samples but slightly less in normal samples, suggesting they might be important in both normal and cancerous processes. **SVIL** and **STARD9** have lower expression in normal samples, which could mean they are upregulated in cancer.




**END**





